File: 1567.乘积为正数的最长子数组长度.py
#
# @lc app=leetcode.cn id=1567 lang=python3
#
# [1567] 乘积为正数的最长子数组长度
#

# @lc code=start
class Solution:
    def getMaxLen(self, nums: List[int]) -> int:
        m = len(nums)
        positive, negative = [0] * m, [0] * m
        if nums[0] > 0:
            positive[0] = 1
        elif nums[0] < 0:
            negative[0] = 1

        maxLength = positive[0]
        for i in range(1, m):
            if nums[i] > 0:
                positive[i] = positive[i - 1] + 1   # 大于正数长度加一
                negative[i] = (negative[i - 1] +
                               1 if negative[i - 1] > 0 else 0)   # 负数同样
            elif nums[i] < 0:
                # 小于 0 时，始终保持 positive 为正数长度，切换 negative 至 positive
                positive[i] = (negative[i - 1] +
                               1 if negative[i - 1] > 0 else 0)   # 未出现负数置为 0
                negative[i] = positive[i - 1] + 1
            else:   # 0 则切分
                positive[i] = negative[i] = 0
            maxLength = max(maxLength, positive[i])  # 始终以 positive 做比较

        return maxLength

# @lc code=end

File: 1567.乘积为正数的最长子数组长度.go
/*
 * @lc app=leetcode.cn id=1567 lang=golang
 *
 * [1567] 乘积为正数的最长子数组长度
 */

// @lc code=start
func getMaxLen(nums []int) (res int) {
	m := len(nums)
	positive, negative := make([]int, m), make([]int, m)
	if nums[0] > 0 {
		positive[0] = 1
	} else if nums[0] < 0 {
		negative[0] = 1
	}
	res = positive[0]
	for i := 1; i < m; i++ {
		if nums[i] > 0 {
			positive[i] = positive[i-1] + 1
			if negative[i-1] > 0 {
				negative[i] = negative[i-1] + 1
			} else {
				negative[i] = 0
			}
		} else if nums[i] < 0 {
			if negative[i-1] > 0 {
				positive[i] = negative[i-1] + 1
			} else {
				positive[i] = 0
			}
			negative[i] = positive[i-1] + 1
		} else {
			positive[i] = 0
			negative[i] = 0
		}
		res = max(res, positive[i])
	}
	return
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

// @lc code=end

